### ISSUE TITLE
Type inference regression with "and" and strict optional disabled

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#11705]
This generates a false positive:

```py
# mypy: no-strict-optional
from typing import Tuple, Dict, Optional, NamedTuple

class C(NamedTuple):
    x: int

t: Optional[C]

d: Dict[C, bytes]

x = t and d[t]
if x:
    x.startswith(b'x')  # Item "C" of "Union[C, bytes]" has no attribute "startswith"
```

It started happening as a side effect of #10389.

This is a regression, so this may be high priority. However, since this only seems to happen when strict optional checking is disabled, the impact may be pretty minor.

cc @ethan-leba

**Comment by sobolevn** [comment]
The problem is that `t and d[t]` is revealed as `Union[Tuple[builtins.int, fallback=ex.C], builtins.bytes*]`. But, it should be revealed as `Optional[Tuple[builtins.int, fallback=ex.C]]`

I will try to solve it.

**Comment by sobolevn** [comment]
Let's test it: https://github.com/python/mypy/pull/11717
